File: 09-backtracking/00090-Subsets_II/00090-subsets_II.cpp
// 90: Subsets Ii
// https://leetcode.com/problems/subsets-ii



#include <iostream>
#include <vector>

using namespace std;

class Solution {
private:
    void backtrack(vector<int>& nums, int start, vector<int>& current, vector<vector<int>>& result) {
        result.push_back(current);

        for (int i=start; i<nums.size(); i++) {
            if (i>start && nums[i] == nums[i-1]) continue;
            current.push_back(nums[i]);
            backtrack(nums, i + 1, current, result);
            current.pop_back();
        }
    }

public:
    // SOLUTOIN
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> current;
        vector<vector<int>> result;
        backtrack(nums, 0, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {1,2,2};

    // OUTPUT
    auto result = o.subsetsWithDup(nums);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 09-backtracking/00090-Subsets_II/00090-subsets_II.go
// 90: Subsets Ii
// https://leetcode.com/problems/subsets-ii

package main

import "fmt"


func backtrack(nums []int, start int, current []int, result *[][]int) {
	*result = append(*result, current)

	for i:=start; i<len(nums); i++ {
		if (i>start && nums[i] == nums[i-1]) {continue}
		current = append(current, nums[i])
		backtrack(nums, (i + 1), current, result)
		current = current[:len(current)-1]
	}
}


// SOLUTION
func subsetsWithDup(nums []int) [][]int {
    var current []int
	var result [][]int
	backtrack(nums, 0, current, &result)
	return result;
}

func main() {
    // INPUT
    nums := []int{1,2,2}

    // OUTPUT
    result := subsetsWithDup(nums)

	fmt.Println(result)
}
